### PR TITLE
Abstract the statement type information out of the match method and into the EndStmtBase subclasses

### PR DESCRIPTION
**Pull request by pelson**
This is a zero functional change which makes it easier to get hold of the data needed to complete #180.

Hopefully this demonstrates a benefit that I hadn't previously made clear to @rupertford (based on https://github.com/stfc/fparser/pull/180#issuecomment-473604008).

Because we are reaching into the class definition, I've changed all of the ``match()`` static methods for EndStmtBase subclasses to be class methods. From a user's perspective, this is an entirely transparent change (you still call ``ClassObject.match(string)``).

I'd like to request that the guideline in https://github.com/stfc/fparser/wiki/CodeReview be relaxed as this is a cross-cutting change that makes sense to be done in one PR (xref https://github.com/stfc/fparser/pull/180#issuecomment-473603065 for context).

### CODE REVIEW

**Comment by rupertford** [comment]
In itself this change is fine and you have addressed one of the problems I had with the version in #180 which specified a `statement_type` string and separately passed the same string to the `match` method, thereby specifying the same information in two places, which is prone to errors and maintenance issues.

However my main issue/question still stands. I do not understand why you need to add a `statement_type` variable when it is already passed into the `match` method and the actual value appears in the `items` list if there is a match. Could you please explain why you think this is required. In #180 the only place it is used is in an error message in `blockbase` and in the `tostr()` method where `items` could (and I believe should) be used.

**Comment by pelson** [comment]
> However my main issue/question still stands. I do not understand why you need to add a statement_type variable when it is already passed into the match method and the actual value appears in the items list if there is a match. Could you please explain why you think this is required. In #180 the only place it is used is in an error message in blockbase and in the tostr() method where items could (and I believe should) be used.

The current codebase has a [``get_type()``](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L1100-L1101) method on an EndStmtBase which currently returns the first parsed item in an EndStmtBase (``"PROGRAM"``, ``"IF"``, etc.). In #180 I am preventing the source from being transformed (to uppercase) and including the ``end`` item in the parse tree so that we can restore the original parsed string with full fidelity (e.g. ``end program`` doesn't become ``END PROGRAM`` upon ``.tostr()``). Therefore, if we want to have ``get_type()`` continuing to work as it did before we need to know what kind of end statement this is, without having to re-interpret the original source. The class *is* enough information to do this ``isinstance(obj, End_Program_Stmt)``, but given we are *already* defining the class type as a string when doing the parsing (as a direct parameter to the ``match`` method) it makes a lot of sense to pull this up to the class' definition for re-use (otherwise we are duplicating this information in disparate code locations).

I'm aware of 4 places that this information is/would-be used:
 * in the match function (as we are seeing in this PR)
 * in the [BlockBase error message](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L558) you referred to (via ``get_type()``) - IMO this *could* be replaced/simplified with behaviour that exists on the EndStmtBase type itself. Ideally the message (and ``get_type()``) would be invariant to the input source, so ``.stmt_type`` would be helpful here.
 * in the [``.tostr()``](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L1117) method as a means of producing code for the parsed source representation - IMO this *could* be replaced with the actual parsed source (as suggested)
 * in the [``.type``](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L1094) attribute on an ``EndStmtBase`` - I have a number of issues with this, as it is ultimately putting redundant data onto the object that can very easily go out of synch (and indeed [already does](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L1110))
   * the ``.type`` attribute is used by [``__repr__``](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L1120) - again, this will move towards using the parsed source rather than the (constant) type name in the repr

In summary, I think ``.stmt_type`` is helpful data that is *invariant* to the parsed source, and can be used in such context as raising consistent error messages (that don't depend on the form/case of the input source). This data *already* exists in the class, I'm just exposing it for re-use rather than have it be duplicated or derived from the source.

**Comment by rupertford** [comment]
So lets concentrate on what you are trying to do.

The first thing is that you want to have more control over the formatting of the output of fparser and are looking at the case of a group of statements. At the moment they are output as upper case. So the natural options I would expect are 1) output the same case as input and 2) have control over the case via some configuration information. Part 2) could be simplified to supporting upper case or lower case to start off with.

Given that we have 1) input text and 2) a matching token, the obvious thing to do is to perform a case insensitive match between 1) and 2) and if there is a match allow the output to be presented as 1) unchanged case and 2) user defined (perhaps lower or upper).

The way to do this in fparser2 would be to do a case insensitive match in the match method, then use the "items" variable to get the case insensitive match value which can then be output as is, or mapped to a particular case (presumably in the tostr()) in some configurable way.

The second point you make is that fparser can add tokens if they are missing and that you have found a bug in this. You suggest that fparser should not output tokens if they are not in the input. What I would suggest here is adding more entries to the "items" output so that we know which tokens exist and which do not. Then it is a simple case of combining the content of the items. This give us case and optional tokens support.

Apologies if I am missing something fundamental but none of the above requires the change you have suggested and as far as I can tell your main argument is that it would make it better for error messages which I disagree with as we would naturally want to output the original text not the matching token.

I think that fundamentally, the only place that the token matching is required is in the match method so I see no reason to make it a class variable for the changes that you are trying to make.

If we reach an impasse then I am happy to pass it by Andy for a second look.

**Comment by pelson** [comment]
Thanks Rupert, helpful information here :+1:

> The second point you make is that fparser can add tokens if they are missing and that you have found a bug in this.

I guess this is the fundamental point. I'm not 100% convinced I think it is a bug. For context (of others reading this as much as anything), the problem is that we have code such as:

```
PROGRAM main
END
```

Being automatically output by fparser as:

```
PROGRAM main
END PROGRAM main
```

fparser clearly knows that ``END`` is an ``EndProgramStmt`` type, but without the context being added by this PR, can't know what the magic word is that should be placed between the ``END`` and the name (``PROGRAM``).

> If we reach an impasse then I am happy to pass it by Andy for a second look.

The impasse only exists if we want ``END PROGRAM main`` behaviour to be preserved I think. I'd happily close this PR in exchange for another that fixes the tests removing the derived components from the output (stmt_type and name), if you believe this is a bug.

> What I would suggest here is adding more entries to the "items" output so that we know which tokens exist and which do not.

I'm reluctant to turn the parse tree into anything more than that because it mandates that we start putting Fortran syntax rules into the logic of things like ``set_name``. I could get behind your "flag" of derived components if we moved towards having the start and end position in the source of each component (so we know exactly where the "name" is defined in the source, to the character) - this information amounts to "derived" vs "not derived" and has other benefits, like being able to provide better context when doing things like static analysis. I haven't dug into the reader and its "item" yet to figure out why this context hasn't already been provided.


In summary, given your feedback, I don't think this PR is going anywhere. I think there are two options I'm comfortable with:
 * prevent ``fparser`` updating and outputting the EndStmtBase name (and stmt_type) unless they were there in the original source
 * start including source locations for items so that we can determine whether an item is "derived" vs "from original source"

Once that discussion has taken its course, happy enough for this PR to be closed out.

**Comment by rupertford** [comment]
> I guess this is the fundamental point. I'm not 100% convinced I think it is a bug. For context (of others reading this as much as anything), the problem is that we have code such as:
> 
> ```
> PROGRAM main
> END
> ```
> Being automatically output by fparser as:
> 
> ```
> PROGRAM main
> END PROGRAM main
> ```
> fparser clearly knows that `END` is an `EndProgramStmt` type, but without the context being added by this PR, can't know what the magic word is that should be placed between the `END` and the name (`PROGRAM`).

Thanks for explaining your concerns @pelson. My take on this is that given `PROGRAM main`, it is equally valid to have `END`, `END PROGRAM`, or `END PROGRAM main` and that the original person writing the code for the match to the `EndProgramStmt` decided not to bother capturing which particular version of the above a particular code uses, rather to just capture that it was an `EndProgramStmt`. As the original form has then been lost they decided to always output the form `END PROGRAM main`.

I don't think any of this should be classed as being magic, they are just optional tokens that the parser arbitrarily decides to always put it irrespective of whether they were all there in the original code.

> The impasse only exists if we want `END PROGRAM main` behaviour to be preserved I think. I'd happily close this PR in exchange for another that fixes the tests removing the derived components from the output (stmt_type and name), if you believe this is a bug.
> 
> > What I would suggest here is adding more entries to the "items" output so that we know which tokens exist and which do not.
> 
> I'm reluctant to turn the parse tree into anything more than that because it mandates that we start putting Fortran syntax rules into the logic of things like `set_name`. I could get behind your "flag" of derived components if we moved towards having the start and end position in the source of each component (so we know exactly where the "name" is defined in the source, to the character) - this information amounts to "derived" vs "not derived" and has other benefits, like being able to provide better context when doing things like static analysis. I haven't dug into the reader and its "item" yet to figure out why this context hasn't already been provided.

At the moment, when the the EndStmtBase match method is called, it is trying to match something of the form:

``` END [ XXX [ YYY ] ] ```

However, it will only ever return two tokens which represent the `END` and the `YYY`. Note, an absent token is returned as `None`. The lines are:

```
            return stmt_type, stmt_name(line)
        return stmt_type, None
```

where `stmt_type` returns the string `END` and the second argument is either `None` if `YYY` does not exist or `YYY` if it does.

What I am saying is that if you want to preserve the input form in the output form then the match method should also return the `XXX` token. It doesn't now because it doesn't matter in terms of generating correct code so the coder did not bother. This would be a relatively simple change ...

```
return token1, token2, token3
```

where `token1` is always the string `end` but it can provide the case of the original. In the case of an EndProgramStmt, token2 is `program` if it exists or None if it does not and token3 is the name if it exists and None if it does not. The additional constraint that would need to be captured is that if token2 is None then token3 should also be None, otherwise it is not a match.

This change would result in the items list of the instantiated class having three values rather than two. Then the tostr() method could output the tokens it received from the input. Does that now make more sense?

> In summary, given your feedback, I don't think this PR is going anywhere. I think there are two options I'm comfortable with:
> 
> * prevent `fparser` updating and outputting the EndStmtBase name (and stmt_type) unless they were there in the original source
> * start including source locations for items so that we can determine whether an item is "derived" vs "from original source"

I believe what I have explained above is your first option. I am very happy for that to be implemented. It can also be used to address the issue of keeping the case of the original code in the output code as all the tokens can be matched in a case insensitive way so the values in the items list have the same case as the input. I'm afraid I don't understand your second option. My opinion is that the only other option would be to leave it as it is :-).

> 
> Once that discussion has taken its course, happy enough for this PR to be closed out.

Hopefully we are getting closer to convergence.

**Comment by pelson** [comment]
> This change would result in the items list of the instantiated class having three values rather than two. Then the tostr() method could output the tokens it received from the input. Does that now make more sense?

I with you on this. The challenge is if ``items`` is ``('END', None, None)``, and then [somebody comes along and asks to set the name](https://github.com/stfc/fparser/blob/0.0.8/src/fparser/two/utils.py#L489). If we are agreeing on bullet point 1 ("prevent fparser updating and outputting the EndStmtBase name (and stmt_type) unless they were there in the original source"), I'll happily follow up with an appropriate PR.

> I'm afraid I don't understand your second option.

>> start including source locations for items so that we can determine whether an item is "derived" vs "from original source"

Let me demonstrate in code:

```
output_elements = []
for element in end_statement.items:
    # The new thing - tells us where in the source this element
    # came from (cursor position & line number)
    if element.source.position is not None:
        # element was in the original source, and wasn't derived after-the-fact
        output_elements.append(element)
    else:
        # Stop. We have reached the first "derived" item, and we only want
        # to output non-derived items. This becomes challenging if we want to
        # be able to manipulate the parse-tree at all (which we do in PSyclone).
        break
```

Essentially, in option 2 I am proposing that we start tracking the position in the line for each element. This is a major piece of work from what I can tell. There are clearly benefits for things like linters and error messages, but for now, option 1 is the quickest and the simplest if we are happy to remove the existing "add the stmt_type and name to the end statement, even if it wasn't in the source" feature.

**Comment by rupertford** [comment]
OK, thanks for explaining. Yes, I think option 1 is the way to go. If you are happy to do that then great!

Regarding your point about setting the name, I think this code should be removed in the associated PR as it is responsible for adding in a name where there was not one before, which is against what we are trying to do in this issue (or whichever issue this idea started in).

**Comment by pelson** [comment]
Option 1 implemented in #192.
